engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/CoerceArrayNode.java

```diff
@@ -0,0 +1,85 @@
+package org.enso.interpreter.node.expression.builtin.mutable;
+
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.data.Vector;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.error.PanicException;
+
+public abstract class CoerceArrayNode extends Node {
+  private @Child InteropLibrary library = InteropLibrary.getFactory().createDispatched(10);
+
+  public static CoerceArrayNode build() {
+    return CoerceArrayNodeGen.create();
+  }
+
+  public abstract Object[] execute(Object value);
+
+  @Specialization
+  Object[] doArray(Array arr) {
+    return arr.getItems();
+  }
+
+  @Specialization
+  Object[] doVector(Vector arr, @Cached HostValueToEnsoNode hostValueToEnsoNode) {
+    try {
+      return convertToArray(arr, hostValueToEnsoNode);
+    } catch (UnsupportedMessageException e) {
+      Builtins builtins = Context.get(this).getBuiltins();
+      Atom err = builtins.error().makeTypeError(builtins.array(), arr, "arr");
+      throw new PanicException(err, this);
+
+    } catch (InvalidArrayIndexException e) {
+      Builtins builtins = Context.get(this).getBuiltins();
+      throw new PanicException(
+          builtins.error().makeInvalidArrayIndexError(arr, e.getInvalidIndex()), this);
+    }
+  }
+
+  @Specialization(guards = "interop.hasArrayElements(arr)")
+  Object[] doArrayLike(
+      Object arr,
+      @CachedLibrary(limit = "5") InteropLibrary interop,
+      @Cached HostValueToEnsoNode hostValueToEnsoNode) {
+    try {
+      return convertToArray(arr, hostValueToEnsoNode);
+    } catch (UnsupportedMessageException e) {
+      Builtins builtins = Context.get(this).getBuiltins();
+      Atom err = builtins.error().makeTypeError(builtins.array(), arr, "arr");
+      throw new PanicException(err, this);
+    } catch (InvalidArrayIndexException e) {
+      Builtins builtins = Context.get(this).getBuiltins();
+      throw new PanicException(
+          builtins.error().makeInvalidArrayIndexError(arr, e.getInvalidIndex()), this);
+    }
+  }
+
+  @ExplodeLoop
```

Comment:
Will exploding loop work? The `argsLength` would have to be a compile time constant then...

Reply:
True